### PR TITLE
Restore alertmanager state from storage as fallback

### LINKED ISSUE #2245: Alertmanager state is not restored from remote storage if replicationFactor == 1
#### Describe the bug
Alertmanager state is not restored from remote storage if replicationFactor == 1: https://github.com/grafana/mimir/blob/main/pkg/alertmanager/state_replication.go#L205

I understand that running replicationFactor at 1 is a bad idea but for dev or staging environments it usually makes sense to keep things simple or cut down on costs.

I was wondering if this is intentional that we skip reading from storage in these cases?

#### To Reproduce

1. Run Alertmanager with replicationFactor of 1
2. Post an alert to AM
3. Kill the AM process
4. Observe state is not restored from storage

#### Expected behavior

State should be read from storage if no replication option is available

#### Environment

- Infrastructure: k8s
- Deployment tool: manual

### PR DESCRIPTION
**Pull request by 56quarters**
Signed-off-by: Nick Pillitteri <nick.pillitteri@grafana.com>

#### What this PR does

In cortexproject/cortex#3925 the ability to restore alertmanager state from
peer alertmanagers was added, short-circuiting if there is only a single
replica of the alertmanager. In cortexproject/cortex#4021 a fallback to read
state from storage was added in case reading from peers failed. However, the
short-circuiting if there is only a single peer was not removed. This has the
effect of never restoring state in an alertmanager if only running a single
replica.

#### Which issue(s) this PR fixes or relates to

Fixes #2245

#### Checklist

- [X] Tests updated
- [na] Documentation added
- [X] `CHANGELOG.md` updated - the order of entries should be `[CHANGE]`, `[FEATURE]`, `[ENHANCEMENT]`, `[BUGFIX]`

### CODE REVIEW

**Comment by callumj** [comment]
Out of curiosity what would be the license implications of backporting this to Cortex?

**Comment by 56quarters** [comment]
> Out of curiosity what would be the license implications of backporting this to Cortex?

I'm not a lawyer so take this with a giant grain of salt:

* I did the fix as a Grafana employee and signed a CLA so they own the copyright.
* The DCO for Cortex requires me to say I have the right to contribute this code and license it under Apache 2 which I don't.
* Cortex could independently fix this based on a bug report similar to the one you submitted to Mimir (excellent report, BTW).

**Review by jhesketh — APPROVED** [review]
lgtm :+1:

File: pkg/alertmanager/state_replication.go [hunk 1]
```diff
@@ -199,38 +199,36 @@ func (s *state) starting(ctx context.Context) error {
...
 		}
-	}
 
-	// The user not being found in all of the replicas is not recorded as a failure, as this is
-	// expected when this is the first replica to come up for a user. Note that it is important
-	// to continue and try to read from the state from remote storage, as the replicas may have
-	// lost state due to an all-replica restart.
-	if err != errAllReplicasUserNotFound {
-		s.fetchReplicaStateFailed.Inc()
+		level.Info(s.logger).Log("msg", "state not settled; trying to read from storage", "err", err)
```
Comment (pracucci) [inline pkg/alertmanager/state_replication.go:229]: I would split this log into 2 logs:
1. Here I should just say that if failed to get the state from other replicas (I would rephrase the log to make it easier to understand)
2. Below, right after the end of the `if s.replicationFactor > 1` conditional block, we add another info log saying we're trying to read the state from storage

Why? Without this change we'll have no info log before going to the storage, but it could be useful to understand what's going on.

File: pkg/alertmanager/state_replication_test.go [hunk 2]
```diff
@@ -161,47 +175,32 @@ func TestStateReplication(t *testing.T) {
...
 				replicator.mtx.Lock()
 				defer replicator.mtx.Unlock()
-				return len(replicator.results) == len(tt.results)
+				return len(replicator.results) == len(tt.replicationResults)
 			}, time.Second, time.Millisecond)
 
 			if tt.replicationFactor > 1 {
+				// If the replication factor is greater than 1, we expect state to be loaded from other Alertmanagers
 				assert.NoError(t, testutil.GatherAndCompare(reg, strings.NewReader(`
-# HELP alertmanager_state_fetch_replica_state_failed_total Number of times we have failed to read and merge the full state from another replica.
```
Comment (pracucci) [inline pkg/alertmanager/state_replication_test.go:169]: Why have been removed? It shouldn't change anything if RF>1.

File: pkg/alertmanager/state_replication_test.go [hunk 3]
```diff
@@ -96,31 +96,39 @@ func newFakeAlertStore() *fakeAlertStore {
...
 	tc := []struct {
-		name              string
-		replicationFactor int
-		message           *clusterpb.Part
-		results           map[string]*clusterpb.Part
+		name               string
+		replicationFactor  int
+		message            *clusterpb.Part
+		replicationResults map[string]clusterpb.Part
+		storeResults       map[string]clusterpb.Part
```
Comment (pracucci) [inline pkg/alertmanager/state_replication_test.go:117]: Why are these maps? Below in the test execution there's `user-1` hardcoded. I think we can simplify it removing the map, given we always test with `user-1`.

**Review by pracucci — APPROVED** [review]
Thanks Nick for working on this! The fix LGTM, but I left few comments I would like you to take a look at, before merging. Thanks!

File: pkg/alertmanager/state_replication_test.go (hunk 2 above)
Comment (56quarters) [reply pkg/alertmanager/state_replication_test.go:169]: I removed this because it was noisy and the most important metric is `alertmanager_state_initial_sync_completed_total` since it more directly represents what this method is testing: what the result of trying to load initial state is.

File: pkg/alertmanager/state_replication_test.go (hunk 3 above)
Comment (56quarters) [reply pkg/alertmanager/state_replication_test.go:117]: Making this change actually makes it a bit more involved due to needing to make them pointers (to represent "no data") and then needing to convert back to values. I'd rather just leave it in this PR if that's alright.

Comment (pracucci) [reply pkg/alertmanager/state_replication_test.go:119]: Ok. I'm fine keeping it as is, given the inconsistency wasn't introduced in this PR. Was nice to apply the boy scout rule, but will be for the next time 😉
